spring-kafka/src/main/java/org/springframework/kafka/requestreply/ValidatingReplyingKafkaTemplate.java

```diff
@@ -0,0 +1,77 @@
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.springframework.kafka.core.ProducerFactory;
+import org.springframework.kafka.listener.GenericMessageListenerContainer;
+import org.springframework.kafka.requestreply.CorrelationKey;
+import org.springframework.kafka.requestreply.ReplyingKafkaTemplate;
+import org.springframework.kafka.support.KafkaHeaders;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * This is a {@link ReplyingKafkaTemplate} that adds a simple validation semantic so it can
+ * take multiple responses and choose the first one that matches the validation condition.
+ * <p>
+ * The use case for this would be providing a farm of topic responders which are decoupled
+ * from the calling service and the calling service does not know who would respond and when,
+ * but knows some property of the response to consider it <em>valid</em>.
+ * <p>
+ * This can be explained using a dinner party analogy:
+ * <ol>
+ * <li>0:00 Kid: Does anyone know what the answers are to the square root of 144 and 2+2?
+ * <li>0:01 Uncle 1: 13, 5
+ * <li>0:02 Uncle 2: 12, 4
+ * <li>0:05 Kid: okay I gathered a few answers, 
+ * <li>0:05 Kid: Filter out who can't answer 2+2
+ * <li>0:05 Kid: The proper answer is 12, 4
+ * <li>0:06 Uncle 3: 12, 4
+ * <li>0:06 Kid: Sorry uncle 3 you're too slow, so I am ignoring you
+ * </ul>
+ */
+public class ValidatingReplyingKafkaTemplate<K, V, R> extends ReplyingKafkaTemplate<K, V, R> {
+
+    private final Function<ConsumerRecord<K, R>, Boolean> validationMethod;
+
+    public ValidatingReplyingKafkaTemplate(ProducerFactory<K, V> producerFactory,
+                                           GenericMessageListenerContainer<K, R> replyContainer,
+                                           Function<ConsumerRecord<K, R>, Boolean> validationMethod) {
+        super(producerFactory, replyContainer);
+        this.validationMethod = validationMethod;
+    }
+
+    public ValidatingReplyingKafkaTemplate(ProducerFactory<K, V> producerFactory, GenericMessageListenerContainer<K, R> replyContainer, boolean autoFlush,
+                                           Function<ConsumerRecord<K, R>, Boolean> validationMethod) {
+        super(producerFactory, replyContainer, autoFlush);
+        this.validationMethod = validationMethod;
+    }
+
+    @Override
+    public void onMessage(List<ConsumerRecord<K, R>> data) {
+        super.onMessage(data.stream()
+                .filter(this::isCorrelationIdMatch)
+                .filter(record -> validationMethod.apply(record))
+                .collect(Collectors.toList()));
+    }
+
+    private boolean isCorrelationIdMatch(ConsumerRecord<K, R> record) {
+        Iterator<Header> iterator = record.headers().iterator();
+        CorrelationKey correlationId = null;
+        while (correlationId == null && iterator.hasNext()) {
```

Comment:
We've changed this on master to use `Header correlationHeader = record.headers().lastHeader(...)`. We should probably do the same here.

Reply:
Would that work on the older version of spring-kafka?

Reply:
Yes (but not with the configurable header name, just the standard header name).

Reply:
Just checked it should workl

Reply:
Yes; making `futures` `protected` will work too but we have to mark it with `// NOSONAR` so our SonarQube checks won't bark.

However, now I think about it, we don't really need a subclass; we can just add a setter on `ReplyingKafkaTemplate`

    setValidationPredicate(BiPredicate<K, R> validationPredicate) {...}

with the default being `(key, value) -> true`.

Reply:
Actually, I think it needs to be `Predicate<ConsumerRecord<K, R>` because the validation might want to look at headers.